examples/pytorch/deepspeed-demo/README.md

```diff
@@ -0,0 +1,37 @@
+### Training a Masked Language Model with PyTorch and DeepSpeed
+
+This folder contains an example of training a Masked Language Model with PyTorch and DeepSpeed.
+
+The python script used to train BERT with pytorch and deepspeed. For more information, please refer to the [DeepSpeedExamples](https://github.com/microsoft/DeepSpeedExamples/blob/master/training/HelloDeepSpeed/README.md).
+
+DeepSpeed can be deployed by different launchers such as torchrun, the deepspeed launcher, or Accelerate.
+See [deepspeed](https://huggingface.co/docs/transformers/main/en/deepspeed?deploy=multi-GPU&pass-config=path+to+file&multinode=torchrun#deployment).
```

Comment:
When using `deepspeed` launcher, it defaults to using pdsh(machines accessible via passwordless SSH)) to send commands to the workers for execution, which is the launcher-worker mode.

The mpi-operator in the training operator is executed through kubectl exec, and it is uncertain whether Deepspeed can support it. Currently, using mpi v2 (via passwordless SSH) would be more appropriate. Deepspeed does not require setting env variables and reads information from the hostfile.
```
# deepspeed --hostfile path default at /job/hostfile
deepspeed --hostfile=/etc/mpi/hostfile /train_bert_ds.py --checkpoint_dir /root/deepspeed_data
```
About hostfile, see: https://github.com/microsoft/DeepSpeed/blob/3b09d945ead6acb15a172e9a379fc3de1f64d2b2/docs/_tutorials/getting-started.md?plain=1#L173-L187
```
# hostfile
worker-1 slots=4
worker-2 slots=4
```

I can add an example in mpi-operator (mpi v2) later.

In PyTorchJob,  `torchrun` and `accelerate` can be used.

Reply:
>The mpi-operator in the training operator is executed through kubectl exec, and it is uncertain whether Deepspeed can support it. Currently, using mpi v2 (via passwordless SSH) would be more appropriate.
 
Thanks for this info! I think, we can support it once we migrate to MPI V2 in TrainJob API. cc @tenzen-y @alculquicondor 
So we can build the specific `deepspeed` runtime that will leverage MPI orchestration to create hostfiles.


>In PyTorchJob, torchrun and accelerate can be used. If I remember correctly, the environment variables for torchrun and accelerate are similar.

As far as I know, the accelerate is compatible with torchrun. However, it might have some additional parameters that `torchrun` doesn't allow to be set. E.g. mixed precision: https://huggingface.co/docs/accelerate/en/basic_tutorials/launch#:~:text=MIXED_PRECISION%3D%22fp16%22

Reply:
`deepspeed` is already compatible with mpi-operator (the one outside of training-operator)

Someone started a PR to add an example, but they abandoned it https://github.com/kubeflow/mpi-operator/pull/610

Reply:
> `deepspeed` is already compatible with mpi-operator (the one outside of training-operator)
> 
> Someone started a PR to add an example, but they abandoned it [kubeflow/mpi-operator#610](https://github.com/kubeflow/mpi-operator/pull/610)

Yes, The image used in this example is one I build earlier. I can provide the Dockerfile for reference. cc @alculquicondor @kuizhiqing

Reply:
I'm happy to accept a PR for this in the mpi-operator repo.

Reply:
I think, once we merge this PR we can refer this training script in MPI-Operator repo as well and add the simple YAML with MPIJob

Reply:
@Syulin7 Yes, thinks to your original work on the base image. The plan in https://github.com/kubeflow/mpi-operator/pull/610 is somewhat staled for some reason. You are really welcomed to continue that.